File: src/p0008/java/src/Solution.java
class Solution {
    public int myAtoi(String s) {

        int i = 0;

        while (i < s.length() && s.charAt(i) == ' ') i++;

        int sign = 1;
        if (i < s.length()) {
            switch (s.charAt(i)) {
                case '-': sign = -1;
                case '+': i++; break;
            }
        }

        int y = 0;
        while (i < s.length() && Character.isDigit(s.charAt(i))) {
            if (y > Integer.MAX_VALUE / 10 || (y == Integer.MAX_VALUE / 10 && s.charAt(i) > '7')) {
                return (sign > 0) ? Integer.MAX_VALUE : Integer.MIN_VALUE;
            }
            y = y * 10 + s.charAt(i) - '0'; i++;
        }

        return sign * y;
    }
}

File: src/p0008/golang/solution.go
package main

import (
	"math"
)

func myAtoi(s string) int {

	i := 0

	for i < len(s) && s[i] == ' ' {
		i++
	}

	sign := int32(1)
	if i < len(s) {
		switch s[i] {
		case '-': sign = -1; i++
		case '+': i++
		}
	}

	y := int32(0)
	for i < len(s) && '0' <= s[i] && s[i] <= '9' {
		if y > math.MaxInt32 / 10 || y == math.MaxInt32 / 10 && s[i] > '7' {
			if sign > 0 {
				return math.MaxInt32
			} else {
				return math.MinInt32
			}
		}
		y = y * 10 + int32(s[i] - '0')
		i++
	}

	return int(sign * y)
}
